File: PDOO/p4/civitas_rb/lib/mazo_sorpresas.rb
require_relative 'sorpresa'
require_relative 'diario'

module Civitas

class Mazo_sorpresas
  
  def initialize
    @sorpresas=Array.new
    @cartasEspeciales=Array.new
    @barajada=false
    @usadas=0
    @ultimaSorpresa = Sorpresa
  end
  
  def al_mazo(sorpresa)
    if @barajada==false
      @sorpresas.push(sorpresa)
    end
  end
  
  def habilitar_carta_especial(s)
    encontrada=false
    for i in  0...@cartasEspeciales.length and !encontrada
      if @cartasEspeciales.at(i)==s
        encontrada=true
        @cartasEspeciales.delete_at(i)
        @sorpresas.push(s)
        Diario.instance.ocurre_evento("carta especial habilitada")
      end
    end
  end
  
  def inhabilitar_carta_especial(sorpresa)
    encontrada=false
    aux= @sorpresas.length
    
    for i in (0...aux)
      if !encontrada
        if @sorpresas.at(i)==sorpresa
        
          encontrada=true
          @sorpresas.delete_at(i)
          @cartasEspeciales.push(sorpresa)
          Diario.instance.ocurre_evento("carta especial inhabilitada")
        
        end
      end
      
    end
  end
  
  def mazo_sorpresas(d)
    @debug=d
    new()
    if d==true
      Diario.instance().ocurre_evento("MazoSorpresas debug true")
    end
  end
  
  def mazo_sorpresas
    new()
    @debug=false
  end
  
  def siguiente
    if ((@barajada==false) or (@usadas==@sorpresas.length))
      if @debug==false
        @sorpresas=@sorpresas.shuffle
        @usadas=0
        @barajada=true
      end
    end
    @usadas=@usadas+1
    @ultimaSorpresa = @sorpresas.at(0)
    @sorpresas.delete_at(0)
    @sorpresas.push(@ultimaSorpresa)
    
    return @ultimaSorpresa
  end
  
  attr_accessor :barajada, :debug, :usadas, :ultimaSorpresa
  
  private :initialize
end
end

File: PDOO/p4/Civitas/src/civitas/MazoSorpresas.java

package civitas;
import java.util.ArrayList;
import java.util.Collections;


public class MazoSorpresas {
    
    private Boolean barajada;
    private Boolean debug;
    private Sorpresa ultimaSorpresa;
    private int usadas;
    
    private ArrayList<Sorpresa> sorpresas;
    private ArrayList<Sorpresa> cartasEspeciales;
    
    
    void alMazo(Sorpresa sorpresa){
        if (barajada == false){
            sorpresas.add(sorpresa);
        }
    }
    
    Sorpresa getUltimaSorpresa(){
        return ultimaSorpresa;
    }
    
    void habilitarCartaEspecial(Sorpresa sorpresa){
        Boolean encontrada = false;
        for (int i = 0; i < cartasEspeciales.size() && !encontrada; i++){
            if (cartasEspeciales.get(i) == sorpresa){
                encontrada = true;
                cartasEspeciales.remove(i);
                sorpresas.add(sorpresa);
                Diario.getInstance().ocurreEvento("carta especial habilitada");
            }
        }
    }
    
    void inhabilitarCartaEspecial(Sorpresa sorpresa){
        Boolean encontrada = false;
        for (int i = 0; i < sorpresas.size() && !encontrada; i++){
            if (sorpresas.get(i) == sorpresa){
                encontrada = true;
                sorpresas.remove(i);
                cartasEspeciales.add(sorpresa);
                Diario.getInstance().ocurreEvento("carta especial inhabilitada");
            }
        }
    }
    
    private void init(){
        sorpresas = new ArrayList<>();
        cartasEspeciales = new ArrayList<>();
        barajada = false;
        usadas = 0;
    }
    
    MazoSorpresas(Boolean debug){
        this.debug = debug;
        init();
        if (debug == true){
            Diario.getInstance().ocurreEvento("MazoSorpresas debug true");
        }
    }
    
    MazoSorpresas(){
        init();
        debug = false;
    }
    
    Sorpresa siguiente(){
        if (barajada == false || usadas == sorpresas.size()){
            if (debug == false){
                Collections.shuffle(sorpresas);
                usadas = 0;
                barajada = true;
            }
        }
        usadas++;
        ultimaSorpresa = sorpresas.get(0);
        sorpresas.remove(0);
        sorpresas.add(ultimaSorpresa);
        
        return ultimaSorpresa;
    }
        
}
